File: datasets/C++/Chatgpt/Algorithm/hard4.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <unordered_set>
#include <algorithm>
using namespace std;

struct State {
    long long val;
    vector<long long> path;
};

int main() {
    long long N;
    int K;
    cout << "Enter N and K: ";
    cin >> N >> K;
    vector<int> A(K);
    cout << "Enter " << K << " elements of A: ";
    for (int i = 0; i < K; i++) cin >> A[i];
    sort(A.begin(), A.end());

    queue<State> q;
    q.push({1, {1}});
    unordered_set<long long> visited;
    visited.insert(1);

    while (!q.empty()) {
        State curr = q.front(); q.pop();
        if (curr.val == N) {
            for (auto v : curr.path) cout << v << " ";
            cout << endl;
            return 0;
        }
        for (int a : A) {
            long long next = curr.val * a;
            if (next <= N && N % next == 0 && !visited.count(next)) {
                visited.insert(next);
                vector<long long> new_path = curr.path;
                new_path.push_back(next);
                q.push({next, new_path});
            }
        }
    }

    cout << -1 << endl;
    return 0;
}

File: datasets/Python/Chatgpt/Algorithm/hard4.py
from collections import deque

def reverse_factorization(N, K, A):
    A.sort()  # For lexicographical order
    queue = deque()
    queue.append((1, [1]))
    visited = set()
    visited.add(1)

    while queue:
        current, path = queue.popleft()
        if current == N:
            return path
        for a in A:
            next_val = current * a
            if next_val <= N and N % next_val == 0 and next_val not in visited:
                visited.add(next_val)
                queue.append((next_val, path + [next_val]))

    return -1

def main():
    N, K = map(int, input("Enter N and K: ").split())
    A = list(map(int, input(f"Enter {K} elements of A: ").split()))
    result = reverse_factorization(N, K, A)
    if result == -1:
        print(-1)
    else:
        print(' '.join(map(str, result)))

main()
